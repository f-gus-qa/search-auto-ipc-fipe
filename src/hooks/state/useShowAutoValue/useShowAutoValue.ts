import _ from 'lodash';
import useSelectState from '../useSelectState';

function useShowAutoValue() {
  const selectedAuto = useSelectState();
  const {
    brand, model, year, value,
  } = selectedAuto;

  const isDone = !Object.values(selectedAuto).includes(null);
  const autoName = isDone
    ? [brand, model, year].map(_.curry(_.get)(_, 'name')).join(' ').trim()
    : null;

  return { isDone, autoName, value };
}

export default useShowAutoValue;
