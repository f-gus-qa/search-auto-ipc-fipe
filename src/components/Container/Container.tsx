import { useEffect, useState } from 'react';
import clsx from 'clsx';
import {
  Container as MuiContainer, makeStyles, Slide,
} from '@material-ui/core';
import useShowAutoValue from '@/hooks/state/useShowAutoValue';
import ShowFipeValue from '../ShowFipeValue';
import FormAuto from '../FormAuto';

const standardTimeTransition = 350;

const useStyles = makeStyles(({ palette }) => ({
  containerForm: {
    minHeight: '100vh',
    display: 'flex',
    justifyContent: 'center',
  },
  main: {
    backgroundColor: palette.success.light,
    minWidth: '100vw',
    minHeight: '100vh',
    transition: `background ${standardTimeTransition + 100}ms ease-in-out`,
    '&.showForm': {
      backgroundColor: palette.primary.light,
    },
  },
}));

function Container() {
  const classes = useStyles();
  const { isDone } = useShowAutoValue();
  const [timeTransition, setTimeTransition] = useState(0);
  const [showFipeValue, setShowFipeValue] = useState(false);
  const [showForm, setShowForm] = useState(!isDone);
  const mainClassName = clsx({
    [classes.main]: true,
    showForm,
  });

  useEffect(() => {
    const timeoutId = setTimeout(
      () => setTimeTransition(standardTimeTransition),
      0,
    );
    return () => {
      clearTimeout(timeoutId);
    };
  }, []);

  useEffect(() => {
    const [setShow, setHide] = isDone
      ? [setShowFipeValue, setShowForm]
      : [setShowForm, setShowFipeValue];

    setHide(false);
    setTimeout(
      () => setShow(true),
      standardTimeTransition + 100,
    );
  }, [isDone]);

  return (
    <div className={mainClassName}>
      <Slide
        direction="down"
        in={showForm}
        mountOnEnter
        unmountOnExit
        timeout={timeTransition}
      >
        <MuiContainer className={classes.containerForm} maxWidth="sm">
          <FormAuto />
        </MuiContainer>
      </Slide>
      <Slide
        direction="up"
        in={showFipeValue}
        mountOnEnter
        unmountOnExit
        timeout={standardTimeTransition}
      >
        <MuiContainer maxWidth="md">
          <ShowFipeValue />
        </MuiContainer>
      </Slide>
    </div>
  );
}

export default Container;

// export default function Container() {
//   return (
//     <div>
//       <div>
//         <div>
//           <div />
//         </div>
//       </div>
//     </div>
//   );
// }
