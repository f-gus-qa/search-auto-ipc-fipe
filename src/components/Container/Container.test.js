import React, { useEffect } from 'react';
import { render, screen, act } from '@testing-library/react';
import { create, act as rendererAct } from 'react-test-renderer';
import {
  fixtureAutoValueComplete,
  fixtureAutoValueIncomplete,
} from '@/fixture/autoValue';
import useShowAutoValue from '@/hooks/state/useShowAutoValue';
import Container from './Container';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useEffect: jest.fn().mockImplementation(
    jest.requireActual('react').useEffect,
  ),
}));
jest.mock('@/hooks/state/useShowAutoValue', () => jest.fn());
jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  Slide: jest.fn().mockImplementation(({ in: inDisplay, children }) => (
    inDisplay ? children : null
  )),
}));
jest.mock('../FormAuto', () => (
  jest.fn().mockImplementation(() => <div data-testid="form" />)
));
jest.mock('../ShowFipeValue', () => (
  jest.fn().mockImplementation(() => <div data-testid="banner-value" />)));

const autoValueComplete = fixtureAutoValueComplete();
const autoValueIncomplete = fixtureAutoValueIncomplete();

describe('<Container>', () => {
  describe('when autoValue is complete', () => {
    beforeEach(() => {
      useShowAutoValue.mockReturnValue(autoValueComplete);
    });

    it('check renders structure', async () => {
      jest.useFakeTimers();
      let toJSON;

      rendererAct(() => {
        toJSON = create(<Container />).toJSON;
        jest.runAllTimers();
      });
      // setTimeout(() => {
      expect(toJSON()).toMatchSnapshot();
      // }, 1000);
    });

    it('check renders banner-value is visible', async () => {
      jest.useFakeTimers();

      act(() => {
        render(<Container />);
        jest.runAllTimers();
      });
      const form = await screen.findByTestId('banner-value');

      expect(form).toBeVisible();
    });
  });

  describe('when autoValue is incomplete', () => {
    beforeEach(() => {
      useShowAutoValue.mockReturnValue(autoValueIncomplete);
      useEffect.mockImplementationOnce(() => jest.fn());
    });

    it('check renders structure', async () => {
      const { toJSON } = await create(<Container />);

      expect(toJSON()).toMatchSnapshot();
    });

    it('check renders form is visible', async () => {
      render(<Container />);
      const form = await screen.findByTestId('form');

      expect(form).toBeVisible();
    });
  });
});
