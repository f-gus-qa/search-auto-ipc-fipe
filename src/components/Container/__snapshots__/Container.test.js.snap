// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container> when autoValue is complete check renders structure 1`] = `
<div
  className="makeStyles-main-2"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthMd"
  >
    <div
      data-testid="banner-value"
    />
  </div>
</div>
`;

exports[`<Container> when autoValue is incomplete check renders structure 1`] = `
<div
  className="makeStyles-main-2 showForm"
>
  <div
    className="MuiContainer-root makeStyles-containerForm-1 MuiContainer-maxWidthSm"
  >
    <div
      data-testid="form"
    />
  </div>
</div>
`;
